website/blog/2024-11-19-ai-lab-first-app.md

```diff
@@ -0,0 +1,220 @@
+
+![a session on the Chatbot recipe](./img/ai-lab-first-app/07-ai-lab-demo-recipe-session.png)
+
+Back to the recipe's details page, we can access the sources of the recipe by clicking on the `Open in VSCode` button, the respository's link or the link `Local clone`.
+
+### Structure of a recipe
+
+The entrypoint of a recipe is the file `ai-lab.yaml` present in the repository of the recipe. 
+
+Let's examine the content of this file (the syntax of the file is specified in [this documentation](https://github.com/containers/podman-desktop-extension-ai-lab/blob/main/PACKAGING-GUIDE.md#recipe-configuration-file)) for the chatbot example.
+
+
+```
+version: v1.0
+application:
+  type: language
+  name: ChatBot_Streamlit
+  description: Chat with a model service in a web frontend.
+  containers:
+    - name: llamacpp-server
+      contextdir: ../../../model_servers/llamacpp_python
+      containerfile: ./base/Containerfile
+      model-service: true
+      backend:
+        - llama-cpp
+      arch:
+        - arm64
+        - amd64
+      ports:
+        - 8001
+      image: quay.io/ai-lab/llamacpp_python:latest
+    - name: streamlit-chat-app
+      contextdir: app
+      containerfile: Containerfile
+      arch:
+        - arm64
+        - amd64
+      ports:
+        - 8501
+      image: quay.io/ai-lab/chatbot:latest
+```
+
+The file defines two containers, one for the inference server and one for the application itself.
+
+The first container, for the inference server, is generic and can be reused for any app using a chat model.
+
+The second one is the one we are particularly interested in. It defines how the container's image for the application is built. It points to the Containerfile used to build the image, on which we can find the source code for the app: in the `app/chatbot_ui.py` file.
+
+Looking at the Python source file, we can see that the application uses the `streamlit` framework for the UI part, and the `langchain` framework for discussing with the model.
+
+We can adapt this source code, by replacing the UI part with a framework to make the app a REST service, and keep the `langchain` part.
+
+An interesting part of the source code is that the recipe does not expose to the user the system prompt, but defines one internally (`You are world class technical advisor`):
+
+```
+prompt = ChatPromptTemplate.from_messages([
+    ("system", "You are world class technical advisor."),
+    MessagesPlaceholder(variable_name="history"),
+    ("user", "{input}")
+])
+```
+
+This is exactly what we want to do in our application, we will be able to indicate here the system prompt we have found earlier.
+
+## Creating our own app
+
+Adapting the source code for the purpose of our application is out of the scope of this article, let's see the result in [our app repository](https://github.com/feloy/podman-ai-lab-demo/blob/main/app/service.py).
```

Comment:
User owned repo, I'm not a big fan, can't we have it as a sub folder of this PR ?

Reply:
The problem had also been raised in the PR https://github.com/redhat-developer/podman-desktop-demo/pull/58

Is it ok if we reference from this blog post the sources in `https://github.com/redhat-developer/podman-desktop-demo`? If it is ok, I'll update the PR above to include the sources of the repo in `https://github.com/redhat-developer/podman-desktop-demo`

https://github.com/redhat-developer/podman-desktop-demo/pull/58

Reply:
Note that if we place the sources in a repo, fetching the recipe from AI Lab will fetch the whole repository. We need to take care of placing it in a reasonably small one

Reply:
Yeah, that's ok to point to the repository on https://github.com/redhat-developer/podman-desktop-demo

Reply:
> Note that if we place the sources in a repo, fetching the recipe from AI Lab will fetch the whole repository. We need to take care of placing it in a reasonably small one

Came to my mind after I posted the comment